File: src/NBdbm/Classes/Fachadas/ParaColl.vb
Imports System
Imports System.Data
Imports System.Collections
Imports System.Globalization

Namespace DotNetDataProviderTemplate
    '*
    '* Because IDataParameterCollection is primarily an IList,
    '* the sample can use an existing class for most of the implementation.
    '*
    Public Class TemplateParameterCollection
        Inherits ArrayList
        Implements IDataParameterCollection

        Default Public Overloads Property Item(ByVal index As String) As Object Implements IDataParameterCollection.Item
            Get
                Return Me(IndexOf(index))
            End Get
            Set(ByVal Value As Object)
                Me(IndexOf(index)) = value
            End Set
        End Property

        Public Overloads Function Contains(ByVal parameterName As String) As Boolean Implements IDataParameterCollection.Contains
            Return (-1 <> IndexOf(parameterName))
        End Function

        Public Overloads Function IndexOf(ByVal parameterName As String) As Integer Implements IDataParameterCollection.IndexOf
            Dim index As Integer = 0
            Dim item As TemplateParameter

            For Each item In Me
                If 0 = _cultureAwareCompare(item.ParameterName, parameterName) Then
                    Return index
                End If
                index += 1
            Next
            Return -1
        End Function

        Public Overloads Sub RemoveAt(ByVal parameterName As String) Implements IDataParameterCollection.RemoveAt
            RemoveAt(IndexOf(parameterName))
        End Sub

        Public Overloads Overrides Function Add(ByVal value As Object) As Integer
            Return Add(CType(value, TemplateParameter))
        End Function

        Public Overloads Function Add(ByVal value As TemplateParameter) As Integer
            If CType(value, TemplateParameter).ParameterName <> Nothing Then
                Return MyBase.Add(value)
            Else
                Throw New ArgumentException("parameter must be named")
            End If
        End Function

        Public Overloads Function Add(ByVal parameterName As String, ByVal type As DbType) As Integer
            Return Add(New TemplateParameter(parameterName, type))
        End Function

        Public Overloads Function Add(ByVal parameterName As String, ByVal value As Object) As Integer
            Return Add(New TemplateParameter(parameterName, value))
        End Function

        Public Overloads Function Add(ByVal parameterName As String, ByVal dbType As DbType, ByVal sourceColumn As String) As Integer
            Return Add(New TemplateParameter(parameterName, dbType, sourceColumn))
        End Function

        Private Function _cultureAwareCompare(ByVal strA As String, ByVal strB As String) As Integer
            Return CultureInfo.CurrentCulture.CompareInfo.Compare(strA, strB, CompareOptions.IgnoreKanaType Or CompareOptions.IgnoreWidth Or CompareOptions.IgnoreCase)
        End Function
    End Class
End Namespace

File: src/NBdbm/CS/Classes/Fachada_Net/ParaColl.cs
using System.Data;
using System.Collections;
using Microsoft.VisualBasic;
using System.Diagnostics;
using System;
using System.Globalization;


namespace NBdbm
{
	namespace DotNetDataProviderTemplate
	{
		//*
		//* Because IDataParameterCollection is primarily an IList,
		//* the sample can use an existing class for most of the implementation.
		//*
		public class TemplateParameterCollection : ArrayList, IDataParameterCollection
		{
			
			
			public object this[string index]
			{
				get
				{
					return this[IndexOf(index)];
				}
				set
				{
					this[IndexOf(index)] = value;
				}
			}
			
			public bool Contains(string parameterName)
			{
				return (- 1 != IndexOf(parameterName));
			}
			
			public int IndexOf(string parameterName)
			{
				int index = 0;
				
				
				foreach (TemplateParameter item in this)
				{
					if (0 == _cultureAwareCompare(item.ParameterName, parameterName))
					{
						return index;
					}
					index++;
				}
				return - 1;
			}
			
			public void RemoveAt(string parameterName)
			{
				RemoveAt(IndexOf(parameterName));
			}
			
			public override int Add(object value)
			{
				return Add((TemplateParameter) value);
			}
			
			public int Add(TemplateParameter value)
			{
				if (value.ParameterName != null)
				{
					return base.Add(value);
				}
				else
				{
					throw (new ArgumentException("parameter must be named"));
				}
			}
			
			public int Add(string parameterName, DbType type)
			{
				return Add(new TemplateParameter(parameterName, type));
			}
			
			public int Add(string parameterName, object value)
			{
				return Add(new TemplateParameter(parameterName, value));
			}
			
			public int Add(string parameterName, DbType dbType, string sourceColumn)
			{
				return Add(new TemplateParameter(parameterName, dbType, sourceColumn));
			}
			
			private int _cultureAwareCompare(string strA, string strB)
			{
				return CultureInfo.CurrentCulture.CompareInfo.Compare(strA, strB, CompareOptions.IgnoreKanaType || CompareOptions.IgnoreWidth || CompareOptions.IgnoreCase);
			}
		}
	}
	
}
